s3gw/handler.go

```diff
@@ -0,0 +1,165 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	s3manager "github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+)
+
+type credentialsProvider struct{}
+
+func (credentialsProvider) Retrieve(ctx context.Context) (aws.Credentials, error) {
+	return awsCredentials, nil
+}
+
+var client *s3.Client
+
+const bucketPathPrefix = len("/bucket/")
+const partSize = 128 << 20 // 128MiB
+
+// getHeaderOrNil gets header value or return nil if the key does not exist
+func getHeaderOrNil(header http.Header, key string) *string {
+	value := header.Get(key)
+	if value == "" {
+		return nil
+	} else {
+		return &value
+	}
+}
+
+// setHeaderOrNil sets header value or delete key if the value is nil
+func setHeaderOrNil(header http.Header, key string, value *string) {
+	if value == nil {
+		header.Del(key)
+	} else {
+		header.Set(key, *value)
+	}
+}
+
+type Object struct {
+	Key  string `json:"key"`
+	Size int64  `json:"size"`
+}
+
+type listResult struct {
+	Objects []Object `json:"objects"`
+}
+
+func newListResult() *listResult {
+	result := listResult{
+		Objects: []Object{},
+	}
+	return &result
+}
+
+func listHandlerFunc(res http.ResponseWriter, req *http.Request) {
+	if req.Method != "GET" {
+		res.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	input := &s3.ListObjectsV2Input{
+		Bucket: &bucketName,
+	}
+	output, err := client.ListObjectsV2(req.Context(), input)
+	if err != nil {
+		res.WriteHeader(http.StatusBadRequest) // XXX we cannot determine appropriate status code
+		res.Write([]byte(err.Error()))
+		return
+	}
+
+	result := newListResult()
+	for _, c := range output.Contents {
+		result.Objects = append(result.Objects, Object{
+			Key:  *c.Key,
+			Size: c.Size,
+		})
+	}
+	marshalled, err := json.Marshal(result)
+	if err != nil {
+		res.WriteHeader(http.StatusInternalServerError)
+		res.Write([]byte(err.Error()))
+		return
+	}
+	res.Header().Set("Content-Length", fmt.Sprint(len(marshalled)))
+	res.Write(marshalled)
+}
+
+func objectHandlerFunc(res http.ResponseWriter, req *http.Request) {
```

Comment:
Let me propose to use a framework rather than implementing routing and response handling on your own. I recommend [Gin](https://github.com/gin-gonic/gin). Sorry if you have already considered this.

Reply:
It's too big for what we want to do.